src/k8s/pkg/k8sd/features/cilium/ingress_test.go

```diff
@@ -0,0 +1,216 @@
+			}
+			g.Expect(status.Enabled).To(Equal(tc.statusEnabled))
+			g.Expect(status.Message).To(Equal(tc.statusMsg))
+			g.Expect(status.Version).To(Equal(cilium.CiliumAgentImageTag))
+			g.Expect(helmM.ApplyCalledWith).To(HaveLen(1))
+
+			callArgs := helmM.ApplyCalledWith[0]
+			g.Expect(callArgs.Chart).To(Equal(cilium.ChartCilium))
+			validateIngressValues(g, callArgs.Values, ingress)
+		})
+	}
+}
+
+func TestIngressRollout(t *testing.T) {
+	t.Run("Error", func(t *testing.T) {
+		g := NewWithT(t)
+
+		helmM := &helmmock.Mock{
+			ApplyChanged: true,
+		}
+		clientset := fake.NewSimpleClientset(
+			&v1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "dummy",
+					Namespace: "kube-system",
+				},
+			},
+		)
+		snapM := &snapmock.Snap{
+			Mock: snapmock.Mock{
+				HelmClient: helmM,
+				KubernetesClient: &kubernetes.Client{
+					Interface: clientset,
+				},
+			},
+		}
+		network := types.Network{}
+		ingress := types.Ingress{
+			Enabled: ptr.To(true),
+		}
+
+		status, err := cilium.ApplyIngress(context.Background(), snapM, ingress, network, nil)
+
+		g.Expect(err).To(HaveOccurred())
+		g.Expect(status.Enabled).To(BeFalse())
+		g.Expect(status.Message).To(Equal(fmt.Sprintf(cilium.IngressDeployFailedMsgTmpl,
+			fmt.Errorf("failed to rollout restart cilium to apply ingress: %v",
+				fmt.Errorf("failed to restart cilium-operator deployment after 3 attempts: %w",
+					fmt.Errorf("failed to restart cilium-operator deployment: %w",
+						fmt.Errorf("failed to get deployment cilium-operator in namespace kube-system: %w",
+							errors.New("deployments.apps \"cilium-operator\" not found"))))),
+		)))
+		g.Expect(status.Version).To(Equal(cilium.CiliumAgentImageTag))
+		g.Expect(helmM.ApplyCalledWith).To(HaveLen(1))
+
+		callArgs := helmM.ApplyCalledWith[0]
+		g.Expect(callArgs.Chart).To(Equal(cilium.ChartCilium))
+		validateIngressValues(g, callArgs.Values, ingress)
+	})
+
+	t.Run("Success", func(t *testing.T) {
+		g := NewWithT(t)
+
+		helmM := &helmmock.Mock{
+			ApplyChanged: true,
+		}
+		clientset := fake.NewSimpleClientset(
+			&v1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "cilium-operator",
+					Namespace: "kube-system",
+				},
+			},
+			&v1.DaemonSet{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "cilium",
+					Namespace: "kube-system",
+				},
+			},
+		)
+		snapM := &snapmock.Snap{
+			Mock: snapmock.Mock{
+				HelmClient: helmM,
+				KubernetesClient: &kubernetes.Client{
+					Interface: clientset,
+				},
+			},
+		}
+		network := types.Network{}
+		ingress := types.Ingress{
+			Enabled: ptr.To(true),
+		}
+
+		status, err := cilium.ApplyIngress(context.Background(), snapM, ingress, network, nil)
+
+		g.Expect(err).To(BeNil())
+		g.Expect(status.Enabled).To(BeTrue())
+		g.Expect(status.Message).To(Equal(cilium.EnabledMsg))
+		g.Expect(status.Version).To(Equal(cilium.CiliumAgentImageTag))
+		g.Expect(helmM.ApplyCalledWith).To(HaveLen(1))
```

Comment:
I'm not a fan of the tight coupling to the implementation details here. There is a time and place for testing the arguments or the number of arguments that a function was called with, e.g. when the function doesn't return anything or isn't easily testable. In this case, we have other information that can be validated, like the status, error, and final state/outcome. Since we can validate the final state or outcome more directly, I'd suggest removing the explicit argument checking from the tests.